designdocs/Capacity.md

```diff
@@ -0,0 +1,657 @@
+
+![https://user-images.githubusercontent.com/3433442/189948939-6b85465a-f9d9-4330-b887-561c7f0283b1.png](https://user-images.githubusercontent.com/3433442/189948939-6b85465a-f9d9-4330-b887-561c7f0283b1.png)
+
+The following interface is implemented on Capacity-Pallet to facilitate replenishment:
+
+### **Hooks**
+
+```rust
+#[pallet::hooks]
+impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+	fn on_initialize(_now: BlockNumberFor<T>) -> Weight {}
+}
+```
+
+Acceptance Criteria are listed below but can evolve:
+
+1. After a fixed number of blocks, a new Epoch Period begins.
+2. At the start of a new Epoch Period,  `EpochNumber` storage is increased by 1.
+3. At the start of a new block, `CurrentBlockUsedCapacity` storage is reset.
+4. At the start of a new block,  `CurrentBlockUsedCapacity` storage is reset.
+5. At the start of a new block,  `CurrentEpochUsedCapacity` storage is incremented with the total Capacity used in the previous block.
+
+### Traits
+
+Replenishable trait implemented on Capacity-Pallet. This trait is used to replenish the Capacity of a  Registered Provider.
+
+```rust
+trait Replenishable {
+  type Balance;
+
+  /// Replenish an MSA amount.
+  fn replenish_by_account(msa_id: MessageSourceId, amount: Balance) -> Result<Balance, DispatchError> {};
+
+  /// Replenish all balance for an MSA.
+  fn replenish_all_for_account(msa_id: MessageSourceId) -> Result<Balance, DispatchError>;
+
+  /// Checks if an account can be replenished.
+  fn can_replenish(msa_id: MessageSourceId) -> bool;
+}
+```
+
+**Storage**
+
+`EpochNumber` help keep count of the number of  Epoch-Periods:
+
+```rust
+/// Storage for keeping count of the number of Epoch periods.
+#[pallet::storage]
+pub type EpochNumber<T> = StorageValue<_, u32, ValueQuery>;
+```
+
+To facilitate keeping track of the amount of Capacity consumed during an Epoch Period.
+
+```rust
+/// Storage to keep accounting for the total Capacity used in an Epoch.
+#[pallet::storage]
+pub type CurrentEpochUsedCapacity<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+```
+
+To facilitate keeping track of the amount of Capacity consumed in a block.
+
+```rust
+/// Storage to keep track of the amount of Capacity consumed in a block.
+#[pallet::storage]
+pub type CurrentBlockUsedCapacity<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+```
+
+**Prioritization of Capacity transactions**
+
+Unlike token transactions, Capacity transactions cannot tip. As a result, this automatically puts Capacity transactions at a disadvantage because it can prevent Capacity transactions from being included in a block in times of high congestion. To prevent token transactions from dominating block space, we prioritize Capacity transactions over token transactions. In addition, we also limit the amount of block space Capacity transactions can consume to prevent these transactions from dominating the block space.
```

Comment:
> we prioritize Capacity transactions over token transactions

This is a bit weird. Tokens seems to have more value since they are non replenish-able. It seems like it should be the other way around